lmcache/blend/retriever.py

```diff
@@ -214,60 +214,73 @@ def __init__(
         self.cache_engine = cache_engine
         self.metadata = metadata
 
-    def _split_input_tokens(self, input_tokens_single_query: torch.Tensor):
-        """Split the input tokens into multiple requests based on the ROI.
-
-        Returns a list of split tokens for cache_engine to retrieve
+    def drop_spt_and_get_indices(
+            self, full_prompt: List[int]) -> Tuple[List[int], List[int]]:
+        """Drop the special token and get the indices of the split requests.
+
+        :param List[int] full_prompt: The full prompts inside one batch.
+        
+        :return: The new prompts without the special token and the indices of
+            the split segments.
+            The indices is recording the start of each segment, ending with
+            the end of the full prompt. 
+            e.g. [0, index_of_segment2, len(full_prompt)]
         """
         spt_len = len(self.spt)
-        if spt_len == 1:
-            indices = (
-                input_tokens_single_query == self.spt).nonzero().squeeze()
-        else:
-            windows = input_tokens_single_query.unfold(0, spt_len, 1)
-            indices = (windows == self.spt).all(dim=1).nonzero().squeeze()
-
-        if indices.dim() == 0:
-            indices = indices.unsqueeze(0)
-
-        start = 0
+        assert spt_len >= 1
+        i = 0
         splitted_tokens = []
-
-        for i in indices:
-            splitted_tokens.append(input_tokens_single_query[start:i +
-                                                             spt_len])
-            start = i + spt_len
-
-        if start < len(input_tokens_single_query):
-            splitted_tokens.append(input_tokens_single_query[start:])
-        return splitted_tokens
+        start = 0
+        while True:
+            next_len = i + spt_len
+            if next_len > len(full_prompt):
+                break
+            if full_prompt[i:next_len] == self.spt:
+                splitted_tokens.append(full_prompt[start:i])
+                start = next_len
+                i = next_len
+            else:
+                i += 1
+
+        if start < len(full_prompt):
+            splitted_tokens.append(full_prompt[start:])
+
+        new_prompt = []
+        new_indices = [0]
+        for split in splitted_tokens:
+            new_prompt.extend(split)
+            new_indices.append(new_indices[-1] + len(split))
+        new_indices[-1] = len(new_prompt)
+        return new_prompt, new_indices
 
     def new_request(
         self,
-        input_tokens: torch.Tensor,
-        query_start_loc: torch.Tensor,
+        full_prompts: List[torch.Tensor],
+        indices: List[List[int]],
```

Comment:
Why are we changing everything from tensor to list?

Reply:
> Why are we changing everything from tensor to list?

For full_prompts, they are just tensors from seq_group_list. So they are organized in a list by vllm before. And if we want to make it one tensor, we need to explicitly concatenate it. I think it is unnecessary. (Note that we cannot use input_ids, the one tensor here for it drops vllm internal cache hit tokens and chunked prefill tokens, if we want that compatibility in the future)

For indices, it is possible to make it List[tensor], but I was just thinking we will probably not cut one prompt into thousands of text chunks.

Reply:
Also, I think even if it is input_tokens, we still loop on the query_start_loc, so the List in full_prompts is not an overhead comparing with before.